agent/src/main/java/io/pyroscope/javaagent/impl/ContinuousProfilingScheduler.java

```diff
@@ -38,14 +39,76 @@ public ContinuousProfilingScheduler(Config config, Exporter exporter, Logger log
 
     @Override
     public void start(Profiler profiler) {
-        Duration firstProfilingDuration;
+        this.logger.log(Logger.Level.DEBUG, "ContinuousProfilingScheduler#start");
+        synchronized (lock) {
+            if (started) {
+                throw new IllegalStateException("already started");
+            }
+            Duration firstProfilingDuration;
+            try {
+                firstProfilingDuration = startFirst(profiler);
+            } catch (Throwable throwable) {
+                stopSchedulerLocked();
+                throw new IllegalStateException(throwable);
+            }
+            this.profiler = profiler;
+            this.executor = Executors.newSingleThreadScheduledExecutor(THREAD_FACTORY);
+            this.job = executor.scheduleAtFixedRate(this::schedulerTick,
+                firstProfilingDuration.toMillis(), config.uploadInterval.toMillis(), TimeUnit.MILLISECONDS);
+            this.started = true;
+        }
+    }
+
+    @Override
+    public void stop() {
+        ScheduledExecutorService svc = this.executor;
+        synchronized (lock) {
+            stopSchedulerLocked();
+            this.executor = null;
+        }
+        if (svc != null) {
```

Comment:
if we need this check here, we probably need it inside `stopSchedulerLocked()` since we access the same reference there.

Reply:
I removed the nil check. I think it is not needed. 
Because of the logic in PyroscopeAgent, if the `scheduler.start()` was not successful or `PyroscopeAgent.Start()` was not called at all, then `scheduler.stop` would never be called, so if we get a call to `scheduler.stop`, then we have an executor
I would appreciate if you could confirm or deny :D

Reply:
I've also removed null check on `job.cancel`